### ISSUE TITLE
TTS voice not always set correctly by OsmAnd's setLanguage()

### ISSUE THREAD

**Issue by tobydickenson** [osmandapp/Osmand#3344]
# Summary

I have found an Android configuration option which unexpectedly makes Osmand UK English TTS sound
a) better, and
b) the same as every other Android TTS app, where previously I have always found it sounds much worse.

# Details

I have always found Osmand's TTS to sound weird. Osmand TTS sounds different to every other Android TTS app. My experience of the "osmand voice" has been relatively consistent over three phones, many factory resets, and countless updates to Osmand and Android, since I first wrote about this problem here in 2013: https://groups.google.com/forum/#!topic/osmand/GdzC2wxQjQE
I have got used to how it sounds, but invariably any guest in my car would ask about the weird voice.

My settings are I think pretty standard for a UK user:
* In Android / Settings / Language & input:
  * Language is: English (United Kingdom)
  * Under Text-to-speech output:
    * Google Text-to-speech engine is selected. Settings are:
      * Language: use system default
      * Under "install voice data", I have "English (United Kingdom) Voice set 1" which was automatically downloaded after factory reset when any app (usually osmand) first uses TTS.
    * The "listen to an example" button says "This is an example of speech synthesis in English" in a good, pleasant, English-sounding voice. The same voice can be heard in every other Android TTS app except Osmand.
* In Osmand:
  * Display language: System (Device language) [and all text is English as expected]
  * Voice guidance: TTS voice English (United Kingdom)

Using 'test voice prompts' option I get a list of only one option: en-gb-tts
I can confirm osmand is using the right voice file using because "attention, pedestrian crosswalk" is spoken as "attention, pedestrian crossing". But it is spoken using a voice which is hard to understand. Definitely *not* the same voice that spoke "This is an example of speech synthesis in English". I have never heard this "osmand voice" from any other Android app.

However, back in Android / Settings / Language & input / Text-to-speech output / Google Text-to-speech engine, what would we expect to happen if I change the language from "use system default" to "English (United Kingdom)"? Note that "English (United Kingdom)" *is* my system language.

As expected, this has no effect on the "This is an example of speech synthesis in English" output. Nor have I have noticed any effect on any other TTS Android app yet. However this change causes Osmand to use the normal Android TTS voice. For the first time in my experience, Osmand voice guidance is easy to understand.

So it looks to me like something in Osmand causes Google TTS to go wrong when it (Google TTS) is configured to "use system default" language.

**Comment by sonora** [comment]
It sounds like a ROM issue. But since you say OsmAnd behaves different from all other apps it is certainly worth investigating. I once created the UK en tts version by simply starting from our default en file, so maybe there is some oversight on my behalf there, not sure. The one thing which is strange is that under "Test voice prompts" you see only en-gb-tts. You should actulally also see all our default languages...

Another oddity is that on all devices I have checked so far the Android tts settings do not seem to offer "System default", you have to explicitly select a language and region (though this may be vendor specific, I checked mostly samsung tts so far). What device and ROM are you actually using? Can you reproduce the issue on another device?

**Comment by tobydickenson** [comment]
I am seeing identical TTS behaviour on:
Moto G, first generation "falcon", running latest stock rom - android 5.1
Wileyfox swift, running latest stock rom - Android 6.0.1
Nexus 4 running latest Cyanogenmod snapshot - Android 6.0.1

All of these have a "system default" tts option - I'll post screenshots next to be more explicit.

Re: only seeing en-gb-tts. I am only seeing this on the nexus4. I think at some point I deleted the "voice" directory, then the normal osmand TTS settings page offers to "install more...", and I only downloaded UK English.

**Comment by tobydickenson** [comment]
![screenshot_20161214-223429](https://cloud.githubusercontent.com/assets/292492/21204167/dd2b0608-c24d-11e6-851d-aeccff8b0a5f.png)
![screenshot_20161214-223436](https://cloud.githubusercontent.com/assets/292492/21204172/e2cba248-c24d-11e6-8b1c-8598b35447d4.png)
![screenshot_20161214-223444](https://cloud.githubusercontent.com/assets/292492/21204174/e5ab8df2-c24d-11e6-855f-87f393f43f0c.png)

**Comment by tobydickenson** [comment]
Which TTS engine are you looking at? All my devices here are using the Google Text-to-speech, which is in the play store. This was the factory default TTS engine on at least the first two devices I listed.

https://play.google.com/store/apps/details?id=com.google.android.tts&hl=en_GB

**Comment by tobydickenson** [comment]
I've just noticed "use system default" is in the screenshot on the play store

**Comment by tobydickenson** [comment]
I can eliminate en-gb-tts as the source of the problem.

Changing Osmand to use en-tts shows exactly the same weird behaviour described above (but with the expected difference that the voice says "pedestrian crosswalk" instead of "pedestrian crossing")

**Comment by sonora** [comment]
Ok. When you use the Pico tts engine for a test, do you see the same behavior?

**Comment by tobydickenson** [comment]
engine | engine setting | osmand tts | result
-|-|-|-
pico | x | fr-tts | french voice
pico | x | en-tts | US English voice
pico | x | en-gb-tts | US English voice
google | x | fr-tts | french voice
google | US English |  en-tts | US English voice
google | US English |  en-gb-tts | US English voice
google | system default |  en-tts | US English voice
google | system default |  en-gb-tts | US English voice (this is what I originally referred to as the "bad" voice)
google | UK English |  en-tts | UK English voice
google | UK English |  en-gb-tts | UK English voice

pico does have a UK English voice, but Ive not found any combination of settings to get Osmand to use it

**Comment by sonora** [comment]
Good table, thanks!

Please note that the OsmAnd language file strictly determines the wording, not the accent, this is expected behavior.

The accent is created by the TTS engine's settings (and quality by their language enhancement packs).

And somehow it seems that Google tts with setting "Use system default" always uses US English, even if your phone is set to UK. And when you have not downloaded the Google tts US English enhancement pack, it defaults to the rudimentary stock version, which procuces the inferior pronunciation and intonation.

Will keep investigating, but chances are there may not be anything in OsnAnd I can do to fix this ...  :(

**Comment by tobydickenson** [comment]
> Please note that the OsmAnd language file strictly determines the wording, not the accent, this is expected behavior

That's not what this table says! The first row of the table says pico gives me a French accent if I choose fr-tts in Osmand, no matter what language setting I have in the pico settings. Row 4 says the same thing about Google TTS.

Many other TTS apps work the way you describe: It is easy to get them to speak French wording in an English accent for comedy results. But not Osmand. In my tests it always uses an accent to match the wording. With the one exception I found here. 

So there must be some code somewhere (either in osmand or android) that *chooses* the voice based on Osmand tts file language code.....

**Comment by tobydickenson** [comment]
> So there must be some code somewhere (either in osmand or android) that chooses the voice based on Osmand tts file language code

TTSCommandPlayerImpl.java calls mTts.setLanguage, then android chooses the voice to match the language.

It looks like osmand is doing nothing wrong....

**Comment by tobydickenson** [comment]
I'm speculating here, because this is my first time with the [android TTS API](https://developer.android.com/reference/android/speech/tts/TextToSpeech.html).....

Osmand calls `setLanguage`, and I'm guessing that Android sees both US English and UK English as the **same** language, even though they have different Locale codes. Osmand doesn't call `setVoice`, which might allow it to switch between a male, female, US, or UK voice.

I'm seeing this breakdown in other TTS apps, for example "Voice Aloud Reader". First choose your language - English - then choose your voice - from a list that includes "Google, UK, Female 1" etc.

Ironically, in my case the result would be much better if Osmand did not call `setLanguage`. I think Android would fall back to the default TTS engine settings. But I agree that's not helpful in general.

**Comment by sonora** [comment]
Yes, from what I remember we may in deed manipulate the tts setting, but likely only set the language, not the region (I had only introduced regional support much later).

Will investigate next week.

**Comment by sonora** [comment]
Still testing here, trying to more completely reproduce it. Current state of my research:
- **(1)** Using any **Samsung TTS engine** on any Samsung device I can find, our current code behaves just fine, as follows: Device locale is "English (United States)": In OsmAnd, selecting en-tts produces the correct en_US accent (high quality, as that Samsung language pack was already downloaded). In OsmAnd selecting en-gb-tts produces a basic quality stock version of an en_GB accent, and the device immediately asks to download the Samsung "English (United Kingdom)" high quality voice file for the Samsung TTS engine.

This proves that the language specification I currently use in our OsmAnd en-gb-tts ttsconfig.p `language('en-gb')` works and also correctly assigns the regional accent (at least on the PROLOG engines of the devices I have tested). (Additional note: I could verify that also both `language('en-GB')`  and `language('en_GB')`  would work the same on these devices).

- **(2)** Now tested the built-in stock Google TTS engines of devices where present. It seems that on those devices where **Google TTS does not offer a language setting "Use system default"**, the en_US or en_GB language selection by merely changing the OsmAnd voice guidance setting (or selection in Test voice prompts) is also handled 100% correctly (equivalent to the test results with the Samsung TTS above), so no issue here. (Looks like this works ok independent of what I set as the Google TTS language setting.)
- **(3)** If, however, I test a device where the **Google TTS offers "Use system language"**, (and  that even seems the preset on these devices), I can reproduce that changing OsmAnd back and forth between en-tts and en-gb-tts does not seem to change the voice (it remains as US accent). I think this is the issue originally reported here.
- **(4)** reading the test table above, there may be a similar issue with **Pico**, but I have not verified. I assume the "x" in the above test table means to say "regardless what setting is used"?

**Comment by OpenSourceForEver** [comment]
The same with spanish:

Device: BQ 5.7 Android 4.4.2
Languaje device: spanish
Google TTS engines: spanish (spain)
Osmand voice: spanish tts

Voice prompts on Osmand are spoken with spanish (USA) but other apps uses spanish (spain).

**Comment by sonora** [comment]
I have added a test button (11.3) at the bottom of the "Test voice prompts" screen showing
- which OsmAnd voice file folder is in use,
- what language selection it suggests in its internal tagging,
- and tapping it shows the system's voice availability status for that tag, and the voice the device's voice engine has chosen as its 'best fit'

We can now play with it. From what I see, it really crucially depends on the TTS engine: While some adjust the voice quite well to the language tag as handed over by OsmAnd, others seem to not respond at all and simply stubbornly use a voice derived by heir own setting.

In other words: Looks like they may not be very good at setting "the current voice to the default one for the given Locale", as it is put in the documentation for `int setLanguage (Locale loc)`:

> Sets the text-to-speech language. The TTS engine will try to use the closest match to the specified language as represented by the Locale, but there is no guarantee that the exact same Locale will be used. (...) This method sets the current voice to the default one for the given Locale; getVoice() can be used to retrieve it.
> 
> 
 

Not sure what can be done about it...

**Comment by tobydickenson** [comment]
> where Google TTS does not offer a language setting "Use system default"

Which version is that? I'm using  3.10.10 of com.google.android.tts which does have "use system default".

> I assume the "x" in the above test table means to say "regardless what setting is used"?

Correct

**Comment by sonora** [comment]
I have replaced "x" by "(any)" in your table for better readability.

Regarding Google TTS versions:
Google Text-to-Speech Engine `3.7.12.2235583.arm.arm_neon` seems to have the "Use system language", and exhibits the bug of not responding to any "setLanguage()" attempt, while `v3.1.3.1162895` seems to always pick a real good fit voice for any OsmAnd voice file you select in OsmAnd.

Will add a few lines to your table to reflect my findings.

**Comment by sonora** [comment]
I have made a small correction (for the case LANG_MISSING_DATA) to our nightly build, which I believe may have fixed (or significantly improved) the original issue.

I have added how it now behaves on my test devices in lines "Google TTS v3.1" and "Google TTS v3.7" of the test table above. Things seem to now simply depend on whether
- a fitting Google voice language is already present on the device
- the device is online or not so it can install one

Could anyone please verify if they have the same findings (a)/(b)/(c) I documented in the table above, and the issue can be considered fixed?

**Comment by sonora** [comment]
In today's nightly build, I have implemented a small change of how OsmAnd sets the Locale for the voice selection.  It will only affect Android 5.0 and higher devices, the method is not available before.

It may produce a more precise Locale specification. Not sure if it helps, from my first tests not, at least not for the Google version I have tried. But this would rule out that the Google engine simply has weaknesses in interpreting the correct language AND country because of a more relaxed specification. Maybe it helps Pico, have not tried.

Please test. Tapping button (11.3) of the dev plugin's "Test voice prompts" screen now produces a popup showing
- which Locale the device tries to set (as interpreted from the language `xx_YY` tag in the ttsconfig.p file)
- what availability message this voice returns when asking the TTS engine to set a voice for it
- and the voice the engine actually uses then.

**Comment by sonora** [comment]
Looks like this may have fixed the issue, at least under Android 5 (?).  I think more cannot be done here.

**Comment by tobydickenson** [comment]
I confirm this has fixed the issue for me

**Comment by sonora** [comment]
Great, thanks for reporting and helping to pinpoint it!  It had gone unnoticed for a long time.  :)

**Comment by OpenSourceForEver** [comment]
It is not fixed with spanish. 
![screenshot_2016-12-23-21-59-50](https://cloud.githubusercontent.com/assets/12786225/21467484/e174b00c-c9ef-11e6-916f-512fcbaacca0.png)

**Comment by sonora** [comment]
@OpenSourceForEver  What Locale is your device set to (in its Android Language settings), and what do you hear, vs. what would you expect?

Please note: In OsmAnd we currently only offer tts config files for "Spanish (Argentina)" (es-ar) and default Spanish (es). For the latter, which variant of Spanish is used should be determined by what default locale your device uses, this is recommended practice. Meaning we currently have no option to force e.g. "Spanish (Spain)" or "Spanish(United States)" (your device makes the selection), it should not be needed (unless someone wants to make wording/vocabulary differences, like for es-ar).

When in OsmAnd you select "Spanish" as TTS voice guidance (or es-tts under the "Test voice prompt", which is the same), that selection does not contain any country/region specification, so I would expect your phone to use the Spanish variant your phone is set to by default (if you have the equivalent TTS voice installed.) That's how it behaves for me in English: When in OsmAnd selecting "en", I receive an American voice when my phone is set to "Enligsh (United States)", and a British voice when it is set to "English (United Kingdom)". By selecting "en-gb-tts" in OsmAnd, I can force it to "English (United Kingdom)" regardless of what the device's default langue is set to, but this will only work if the system has a British TTS language installed. Once I delete that, it would fall back to a US English voice.

I would expect it works in perfect analogy with OsmAnd's Spanish (es) voice, and the Spanish (Argentina) (es-ar) voice we offer?

Please note that the display of what voice is used is somewhat limited under Android 4.4.2. Your screenshot above only shows that your phone selected its default Spanish voice ("spa"), but I cannot tell which one that is. Again the English analogy here: Phones with Android 5+ always report the complete locale used. But my Android 4.4.2 test devices simply report "eng" when referring to American English, and would report "eng_GBR" when I succeeded in selecting a British voice.

**Comment by OpenSourceForEver** [comment]
My phone languaje is spanish (there isn't any variant on device settings). Maybe it depends on Android version.

On text to speech output (device), Spanish (Spain) is selected. If I push "listen to an example" the voice output is Spanish (Spain) but if I use Osmand, voice output is Spanish (USA) (other apps in the same case use Spanish Spain)
If I delete Spanish (USA) on text to speech settings (device) Osmand uses Spanish (Spain) but if Spanish (Spain) and Spanish (USA) are installed on device, Osmand uses Spanish (USA).

The difference between Spain and USA is pronunciation. Vocabulary used is the same as ttsconfig.p (Osmand) on both cases.

**Comment by sonora** [comment]
I have no device here which does not offer "Espanol (Espana)" as a system locale. And all these do not seem to have that issue. Maybe the best general solution for you is you unhide "Espanol (Espana)" on your device using the "More locales" app?  Feliz Navidad!

**Comment by OpenSourceForEver** [comment]
Ok. Thanks for your answer.
